### ISSUE TITLE
Passive tests can fail when -10.pA injections fail to perturbe V_{m}

### ISSUE THREAD

**Issue by russelljjarvis** [russelljjarvis/NeuronunitOpt#32]
Passive tests can fail when -10.pA injections fail to perturbe $V_{m}$ If np.std(vm)==0, then a curve fitting routine in passive tests fails. You can choose to assign a poor score, however, I believe the Druckman protocol tries a sequence of 3 increasingly negative pulses, and I don't think I see the same error. Therefore the Druckman input resistance test is probably more robust.

**Comment by rgerkin** [comment]
This is a little bit crazy, but in the interest of pushing forward what if we created a `RandomScore` which would literally just contain a random `score`, and use that for all otherwise unhandled problems in computing scores.  If the solution contained a `RandomScore` instead of another type, then we could mark it for investigation.  In the meantime at least optimizations would run to completion (the gradient in that direction would be random, but shouldn't systematically break things).  

Maybe even better, we could make the `RandomScore` decrease with every generation, so that the optimizer wouldn't get distracted looking to optimize on things that perturb the `RandomScore`.  

I may be misunderstanding how your code works, so if this can't work let me know.

**Comment by rgerkin** [comment]
More basic question... how is -10 pA failing to perturb `V_m`?  In any cell that amount of current should move Vm`V_m`? by more than machine precision, unless your input resistance is very, very, very close to zero.  Could you give an example of a model and set of parameters where this occurs?

**Comment by russelljjarvis** [comment]
Sure, yes its an error I see occassionally when running the GA with different models backends. It might only happen when sampling extreme parameter values.

What if the random score is more like a "flag-suspicious" score, and instead of being random, it is systematically uninteresting to the optimizer, like the mean of someother scores in that generation (not garunteed exist)

**Comment by rgerkin** [comment]
If you want to flag suspicious, you could just create a `FloatScore` (already a score type), but use the mean of the others like you said.  The fact that you aren't using `FloatScore` for anything else should already flag it, but you could also add an extra attribute at run-time if needed.

**Comment by russelljjarvis** [comment]
I am running out of time to implement something else, or create minimal working examples. My current solution is to wrap passive tests in a try-catch exception handling. If test fails, then a bad score is returned.

**Comment by rgerkin** [comment]
That's fine, but I would like some sort of bread-crumb trail to identify these later.  Like if the bad score is bad enough that it couldn't possible be caused by anything but an error, then that would help highlight the bug for future fixing.

**Comment by russelljjarvis** [comment]
In the case of the Hodgkin Huxley models,  an entire first generation of genes fail passive tests for this reason. It could be more a problem with knowing suitable model parameters.

**Comment by rgerkin** [comment]
Point me to some code to instantiate an HH model with a given set of parameters, run, and extract a Vm trace, and I will report back with parameter ranges that you should use.

**Comment by russelljjarvis** [comment]
If you run notebook chapter10.ipynb.  approximately cell 4, or 5. Set MU=NGEN=10 (to guarantee a wide exploration of Param-space)
 Edit  [code here](https://github.com/russelljjarvis/neuronunit_opt/blob/master/neuronunit/optimisation/optimization_management.py#L1881-L1885).
 to erase exception handling wrapper.

**Comment by rgerkin** [comment]
What I am looking for is something like this in the canonical example form:
```
from neuronunit.models import SomethingModel
model = SomethingModel(backend='bhh')
model.attrs = {'C': 1.5e-06,
               'EK': -0.018,
               'ENa': 0.0575,
               ...}
vm = model.get_membrane_potential()
```

Except I don't know what model class to use (I made up `SomethingModel` just now).  I cannot piece this together from the chapter10 notebook because they are so many layers of abstraction between optimization and running a model with a desired set of parameters.  

So what would the minimal example for generating a membrane potential from a given set of HH parameters look like, following the template above?

**Comment by russelljjarvis** [comment]
AOK that should be easy. `ReducedModel` works.

**Comment by russelljjarvis** [comment]
```python
from neuronunit.optimisations.optimization_management import new_model
from neuronunit.optimisation.data_transport_container import DataTC as pre_model                  
HH_params =  { 'El' : 10.6 * b2.units.mV,
        'EK' : -12 * b2.mV,
        'ENa' : 115 * b2.mV,
        'gl' : 0.3 * b2.msiemens,
        'gK' : 36 * b2.msiemens,
        'gNa' : 120 * b2.msiemens,
        'C' : 1 * b2.ufarad,
	'Vr':-80.0 }
pre_model.attrs = HH_params
pre_model.backend = str("BHH")  
model = new_model(pre_model)
```

**Comment by russelljjarvis** [comment]
If you prefer some other kind of syntax, please feel free to make a method and introduce it into the optimization_management.py file.

**Comment by russelljjarvis** [comment]
![image](https://user-images.githubusercontent.com/7786645/69488579-553fac00-0ebf-11ea-87b7-756ba75b7c73.png)

**Comment by russelljjarvis** [comment]
this method is now(sih) in optimization_management
``` python
from neuronunit.optimisation.optimization_management import inject_and_plot
attrs = { 'El' : 10.6,
        'EK' : -12,
        'ENa' : 115,
        'gl' : 0.3,
        'gK' : 36,
        'gNa' : 120,
        'C' : 1,
	'Vr':-55.0 }
backend = str("BHH")  
vm = inject_and_plot(attrs,backend)
```

**Comment by rgerkin** [comment]
OK, I will use that.  

And so I understand this correctly, I think your approach is that you start with a that `DTC` (model parameters, etc.), use it to create a model from which you get a rheobase (this being the first step to many things), which is stored in that `DTC` (or a copy thereof).  Then you create a fresh model, which is basically always instantiated with the Izhikevich LEMS file, although you never use any of the information in that file -- it's just an easy path to having a NeuronUnit model.  Then you copy the DTC's attributes to that model and run simulations.  

I should probably create a more basic model than `LEMSModel` (which `ReducedModel` subclasses) which still uses backends to implement things like simulation and getting the membrane potential, but is completely agnostic to the form of the original model (e.g. a LEMS file, a PyNN model, a dictionary of HH parameters, etc.)  If you imagine a simple example of creating a model and optimizing it, it would be sort of silly to instantiate based on a LEMS file that is never used (e.g. the LEMS file is Izhikevich but the optimization is really for a HH model with HH attributes), but I think that is what it would currently look like without such a more fundamental class.

**Comment by russelljjarvis** [comment]
This is all really accurate.

I took a big stab at implementing a class VeryReducedModel, its remants are probably in a file called that.

It doesn't use LEMS at all, but it yielded cryptic error messages.

**Comment by russelljjarvis** [comment]
It might be that the data_transport_container object almost is a model, and may as well be elevated to the status of a model, when appropriate by binding methods to the class that are required by sciunit.

Like for instance in the data_transport_container class there is a method 

Maybe all that is necessary is to bind the sciunit model methods to this class to make it amenable to sciunit.generate_prediction/exctract_feature/compute_score/judge etc.
```
def dtc_to_model(self):
   return sciunit_model
```

We could think of such a model is as a class that is not designed to persist, or propagate through algorithms, but something that is used in-place to compute scores as needed, and then disposed of/garbage collected.

**Comment by russelljjarvis** [comment]
One interesting thing to keep in mind is that, although historically NeuronUnit Models were not pickable, and transmissable between CPUs, this is not something I have continued to check up on every 6 months. Maybe they are portable now.

Although the DataTC class is constrained in many ways to garuntee pickling/transmission, there is nothing preventing self methods in the class from outputting unpicklable models.

I actually think this would be the best future direction for program design, is to have a method inside DataTC, that repackages it `self` as an unpicklable NU model.

**Comment by rgerkin** [comment]
Right, it is possible that DTC is obsolete (since models may now be pickable) but I am also happy to keep them around because at the very least they will probably always be a more robust, memory-efficient vessel for the information that is essential to optimization.  The model itself will always have a bunch of things that are a waste of compute cycles to serialize/deserialize and pass around.  

So I agree that something like `Model.to_dtc()` and `DTC.to_model()` are a good route.  I see that `DTC.dtc_to_model()` already exists.  I think we can leave that as it is for now.  I can make a `Model.to_dtc()`, unless there already is one (or something equivalent).

**Comment by russelljjarvis** [comment]
Yes Model.to_dtc is a good idea. I don't think I have that. Although I have `DTC.dtc_to_gene()` in the data_transport_container file.

Also in optimisation/algorithms.py there is a function called `strip_object` that is used to shed unpickable object attributes. I think that came from your backend code somewhere.

**Comment by russelljjarvis** [comment]
For the minimal working example, that I am still resistant to providing in issue [20](https://github.com/russelljjarvis/neuronunit_opt/issues/20).

If you really want to go down such a route, you can add to the example below.

``` python
from neuronunit.optimisation.optimization_management import inject_and_plot_model
attrs = { 'El' : 10.6,
        'EK' : -12,
        'ENa' : 115,
        'gl' : 0.3,
        'gK' : 36,
        'gNa' : 120,
        'C' : 1,
	'Vr':-55.0 }
vm = inject_and_plot(attrs,str("BHH"))
```

Then do.
```import os
electro_path = str(os.getcwd())+'/../tests/russell_tests.p'

assert os.path.isfile(electro_path) == True
with open(electro_path,'rb') as f:
    (test_frame,obs_frame) = pickle.load(f)
local_tests = test_frame['Hippocampus CA1 pyramidal cell']
```
```
dtc.tests = local_tests
dtc.format_test()
model = dtc.dtc_to_model()
for t in dtc.tests:
    model.judge(t)
```

**Comment by russelljjarvis** [comment]
I have since refactored code such that I use the file:
reduced_model_sans_lems, that inherits from sciunit.model base, and neglects to inherit from LEMS.

I now use the syntax 
```python
model = dtc.dtc_to_model() 
```
exclusively, and 
```
model = mint_generic_model(backend) 
```
Will not work. My notebooks will mostly be broken until I update them all.

**Comment by russelljjarvis** [comment]
It turns out that on my branch passive, the test parameters for current injection values had a default value of ```-0.0 pA``` that was not subsequently being over written for some reason.

**Comment by rgerkin** [comment]
OK, awesome.  Can I close this issue then?

**Comment by russelljjarvis** [comment]
Yes.

**Comment by russelljjarvis** [comment]
Actually with protocols updating the issue still persists.

**Comment by rgerkin** [comment]
@russelljjarvis Do you mean the issue returned when protocols were changed in the code, or that 0 pA is injected when the protocol changes within one optimizer run, or something else?

**Comment by russelljjarvis** [comment]
"you mean the issue returned when protocols were changed in the code". Yes the issue persisted.

**Comment by russelljjarvis** [comment]
I no that the negative current injection is being applied, as I can see appropriate negative deflections in voltage traces, in the ascii plots.

**Comment by rgerkin** [comment]
So you can see the negative deflection of Vm, but as far as you can tell the stdev (over a region including baseline and the deflection) is still computed to be 0?

**Comment by russelljjarvis** [comment]
I believe so, yes. It's possible there is something else going on. I can wrap the expression in a `try`, `except` block, and pickle the model parameters that raise this kind of exception, so they can be invoked one by one in a notebook.

**Comment by russelljjarvis** [comment]
Maybe its just the units here units are stripped:
```python
    @classmethod
    def get_segment(cls, vm, start, finish):
        start = int((start/vm.sampling_period).simplified)
        finish = int((finish/vm.sampling_period).simplified)
        return vm[start:finish]
```
Later they are put back in:
```python
    @classmethod
    def exponential_fit(cls, segment, offset):
        t = segment.times.rescale('ms')
        start = t[0]
        offset = offset-start
        t = t-start
        t = t.magnitude
        vm = segment.rescale('mV').magnitude
```

**Comment by rgerkin** [comment]
The units being stripped are units of time.  I think there might be a way to not do this (to simply slice the segment by time points instead of indices) but I'm not sure how the time stripping would interact with the voltage scale issue.

**Comment by russelljjarvis** [comment]
At the moment I have a file backend_tests. This file/code calculates rheobase on a standard model/backend combination, injects rheobase current and plots it.

I can extend the same file to inject and plot ```-10*pq.pA``` and eyeball the waveforms, with theirir units.

**Comment by rgerkin** [comment]
We need a function to consistently rescale all quantities to MKS units (or something like that) before stripping units (and then a companion function to undo this stripping of units), so that unitless variables always have the correct implied units.

**Comment by rgerkin** [comment]
Also I need to look into numerical precision issues with exponential fits.  This might mean using a unit other than Amps, for example, to avoid very small numbers near the floating point boundaries.

**Comment by rgerkin** [comment]
We could compute the mean of the waveform and then divide by it before computing the time constant.

**Comment by rgerkin** [comment]
Implemented in #41.
